### PR TITLE
[GOBBLIN-1588] Send failure events for write failures when watermark is advanced in MCE writer

### PR DESCRIPTION
**Pull request by jack-moseley**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1588

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Sent failure events in both cases where a watermark could be advanced
1. After a flush due to operation change, if a previous write failed
2. After a periodic flush where GMCE watermark will be advanced

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Updated unit test to check that events are sent

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by jack-moseley** [comment]
@ZihanLi58 please review

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java [hunk 1]
```diff
@@ -420,4 +434,22 @@ public static State setHiveRegProperties(State state, GobblinMetadataChangeEvent
...
     return tmpState;
   }
+
+  /**
+   * Submit event indicating that a specific set of GMCEs have been skipped, so there is a gap in the registration
+   */
+  private void submitFailureEvent(GobblinMetadataException exception) {
+    log.warn(String.format("Sending GTE to indicate table flush failure for %s", exception.tableName));
+
+    GobblinEventBuilder gobblinTrackingEvent = new GobblinEventBuilder(IcebergMCEMetadataKeys.ICEBERG_COMMIT_EVENT_NAME);
```
Comment (ZihanLi58) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java:444]: Do we want to use the same event name when we successfully commit one snapshot? Should we use another event name like METADATA_WRITER_FAILURE_EVENT?

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java [hunk 2]
```diff
@@ -420,4 +434,22 @@ public static State setHiveRegProperties(State state, GobblinMetadataChangeEvent
     }
     return tmpState;
   }
+
+  /**
+   * Submit event indicating that a specific set of GMCEs have been skipped, so there is a gap in the registration
+   */
+  private void submitFailureEvent(GobblinMetadataException exception) {
+    log.warn(String.format("Sending GTE to indicate table flush failure for %s", exception.tableName));
```
Comment (ZihanLi58) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java:442]: also include the db information?

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java (hunk 1 above)
Comment (jack-moseley) [reply gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java]: Whoops yeah I meant to change that but forgot

**Pull request merged**